soapbox/python/nicolascarbone.md

```diff
@@ -0,0 +1,145 @@
+
+<br>
+
+### Private Class Properties
+
+Python does not have access modifiers so use a leading `_` for private methods.
+
+Avoid using leading `__` because that will trigger a process called `name mangling`, it will just put the classname as attribute preffix.
+
+## Coding Practices
+
+As general rule, your code should have high cohesion and low coupling, this will make it easier to maintain, scale, etc.
+
+Some things to keep in mind:
+- Dependency Injection: This will decouple your services from other ones and it will make them easier to test and maintain.
+
+    See [Example](https://gist.github.com/nicolascarbone/48a837ecab8b8d84250deaf012b36307)
+    and this [post](https://levelup.gitconnected.com/dependency-injection-in-python-3-with-testing-55b64ca60ae1) for a
+    more complete example.
+
+- Repository pattern: all CRUD operations should go here so your services won't get tied to the db layer.
+
+    See [Example](https://gist.github.com/nicolascarbone/4838a48852b1ae03b8095dc4cfa8d008)
+
+    Then you can use this class to access to your data from your services.
+
+- Single responsability: Avoid adding more than one responsability to each function/method. This will make you code simpler and easier to test.
+- Unit of work: this is a useful pattern which collaborates with the repository pattern to coordinate the writing out of changes. It is an abstraction for atomic operation, so the API does not communicate with all the other layers.
+
+    See [Chapter](https://www.cosmicpython.com/book/chapter_06_uow.html) of the amazing book by Harry Percival.
+
+### Accessing to the database.
+
+Whenever possible use an ORM to access to the database, this can help to make your app compatible with different db engines like:
+- Postgres
+- Mssql
+- Sqlite
+
+A common option is [SqlAlchemy](https://www.sqlalchemy.org) but there are others like [Peewee](http://docs.peewee-orm.com/en/latest/).
+
+When using an ORM avoid writing raw sql unless is strickly necessary, this will tie your code to 
+an specific db engine.
+
+### Testing
+
+Python has unittest as part of the standard lib, however, there are more flexible tools and frameworks
+like [Pytest](https://docs.pytest.org/en/stable/index.html), which has some really useful features that makes
+tests more scalable and it is widely accepted by the python community.
+
+Considerations:
+- Each function or method should have its own test
+- Keep an eye on responsabilites and make sure you function/method does only one thing,
+    this will allow you to write smaller and more reliable tests.
+- Mock any third-party lib.
+    - You can use mock and patch from unittest for this purpose.
+- Use dependency injection: this will allow you to use mocks to easily test services without patching a lot of different classes.
+
+    ```
+    cache = mock.Mock()
+    my_service(cache=cache)
+    cache.get.assert_called_with(key)
+    ```
+
+### Frameworks
+
+    <link to fastapi soapbox> ?
+
+### Virtual Envs
```

Comment:
I feel like these three sections should be up top... Maybe it's just me but this is the type of stuff I'd be looking for _first_! I always struggle with this crap in python

Reply:
LOL!
that makes sense
I moved the environment section to the top